repos/system_upgrade/el7toel8/models/pcidevices.py

```diff
@@ -17,9 +57,24 @@ class PCIDevice(Model):
     driver = fields.Nullable(fields.String())
     modules = fields.Nullable(fields.List(fields.String()))
     numa_node = fields.Nullable(fields.String())
+    pci_id = fields.String()
 
 
 class PCIDevices(Model):
     topic = SystemInfoTopic
 
     devices = fields.List(fields.Model(PCIDevice))
+
+
+class RestrictedPCIDevices(Model):
+    """Data about restricted (unsupported or unavailable) PCI devices.
+
+    driver_names - is a container with the devices, which are restricted by
+        the driver name identifier
+    pci_ids - is a container with the devices, which are restricted by
+        the pci_id identifier
+    """
+    topic = SystemInfoTopic
+
+    driver_names = fields.JSON()
+    pci_ids = fields.JSON()
```

Comment:
Add docstrings for fields in models as well,.

Reply:
I didn't get this. There are docstings added to this new model and to PCIDevice. What else I can do?

Reply:
If you meant PCIDevices, then, I don't think it is necessary

Reply:
Ah. I see. I missed that the info is actually part of the model dosctring. Well, I am more thinking about something like this:
```python
class RestrictedPCIDevices(Model):
    """
    Data about restricted (unsupported or unavailable) PCI devices.
    """
    topic = SystemInfoTopic

    """
    Container with devices restricted by the driver name identifier
    """
    driver_names = fields.JSON()

    """
    Container with devices restricted by the pci_id identifier
    """
    pci_ids = fields.JSON()
```

But so far as everything is describe in the docstring, I think it's ok as well.

Reply:
Taking back. currently I am not able to discover what it the content of the field without looking into the code - unit-tests. Why not to represent the info by model so it's transparent? I think that we should not use `fields.JSON()` when it's possible.

Reply:
@pirat89 Any ideas on how to actually improve it? 
```python
from typing import Dict, TypedDict, Optional


class RestrictedPCIDevice(TypedDict):
    pci_id: str
    driver_name: Optional[str]
    device_name: Optional[str]
    available_rhel7: int
    supported_rhel7: int
    available_rhel8: int
    supported_rhel8: int
    available_rhel9: int
    supported_rhel9: int
    comment: Optional[str]


class RestrictedPCIDevices(TypedDict):
    driver_names: Dict[str, RestrictedPCIDevice]
    pci_ids: Dict[str, RestrictedPCIDevice]


# As a leapp Model definition
class RestrictedPCIDevices(Model):
    """Data about restricted (unsupported or unavailable) PCI devices.

    driver_names - is a container with the devices, which are restricted by
        the driver name identifier
    pci_ids - is a container with the devices, which are restricted by
        the pci_id identifier
    """
    topic = SystemInfoTopic

    driver_names = fields.JSON()
    pci_ids = fields.JSON()
```

Reply:
The point is to not have dict or something like that in models (which JSON is kind of like dict. we do not want it). Solution is e.g.
```python
from leapp.models import Model, fields
class RestrictedPCIDevice(Model):
    pci_id = fields.String()
    driver_name = fields.Nullable(fields.String())
    ....

class  RestrictedPCIDevices(Model):
    driver_names = fields.List(fields.Model(RestrictedPCIDevice))
    pci_ids = fields.List(fields.Model(RestrictedPCIDevice))

```

Reply:
It is a very different data structures. I need some associative array (Mapping data strcture) to represent the restricted driver names and pci_ids. So the search through them will be O(1)

Reply:
I think that even with O(n) search we will not be able to occur any performance issues. You can check the rest of actors. We have messages that have tousands of items and are checked almost in every actor and the performance impact is minimal. you can create just functions to get desired data based on a value and do not have to transform it to associative arrays. or you can build new objects to represent it in associative way. But regarding the point what is the most time consuming stuff during IPU, you will not really realize the difference. we are talking here about couple of ms in the worst realistic scenrario.

Reply:
+1 to Petr here, we are talking about 1000 entries tops.

Reply:
My point is that Dict is totally valid data structure and it has its applications. Let's see what you're suggesting (simplified):
```
Dict[str, Model] (comes from mcs) -> List[Model]. Then we have to loop all drivers on host with all drivers in List[Model] to check if there is a match. Or construct the Dict[str, Model] inside actor and do the same like I have it now. 
```
What I am suggesting (simplified):
```
Dict[str, Model] (comes from mcs) -> Dict[str, Model] (analogue using leapp Model) -> then we get the required Model just by accesing the Dict[str, Model]
```
+ create a ticket to add additional data type, supporting the validation of dicts (it is actually not that hard).

Reply:
I see no problem with lists but dicts in a model is a PITA. Maybe the team sees things differently, but I would strongly prefer list here.

Reply:
@fernflower Inna, could you please:
1. Expand what is PITA
2. Explain why do you "strongly prefer list here"?

Reply:
As I pointed on the stars, you cannot validate data in dicts from the nature of dicts. That's the pita. It is not possible validate what "keys" you have in the dict. That's the reason why dicts are forbidden to be used in models in the project. It's reason why you do not see dicts in any model in the project and why you see lists. From the nature of model design we use, never-ever is expected to happen that you ask for `foo[bar]` and get the "KeyError" traceback. Basically you cannot verify, what all data you have inside with dict. That's why you should always provide models which are consistent (dict is inconsistent from the nature of dict in this case).

Reply:
The goal of validation is to check the data types, not the data values. And you can validate data types of all keys and all values.

Anyway, it will be better to talk about this o our cabal session. Thanks for pyutting your efforts here

Reply:
I will try to put it in different way. In current models, when you have
```
class ModelA(Model):
    myfield = fields.String()
```

I can be very sure, that the field is there everytime. As well, I could have:
```
class ModelA(Model):
    myfield = fields.Nullable(fields.String())
```
and I can be very same sure, there is the field with string or None object. I can be sure about data type(s). But in the case of dict, you are not able to tell anything like that. Because the fields become to be dynamically created strings. So you cannot do the validation. You could have there whatever (e.g.) strings. So I cannot be sure I ask modelA["astring"] whether is there or not. And that's the thing, when the value itself becomes part of data. Because instead of data, I could get simply traceback.

Not sure whether this explanation helps or not. Anyway, keeping that for tomorrow's cabal session.

Reply:
@pirat89 I understand what you're describing. 
I did a little update to the model:
```python
class RestrictedPCIDevices(Model):
    """
    Data about restricted (unsupported or unavailable) PCI devices.

    driver_names - is a container with the devices, which are restricted by
        the driver name identifier
    pci_ids - is a container with the devices, which are restricted by
        the pci_id identifier
    """
    topic = SystemInfoTopic

    driver_names = fields.Dict(
        fields.String(),
        fields.Model(RestrictedPCIDevice),
        default={},
    )
    pci_ids = fields.Dict(
        fields.String(),
        fields.Model(RestrictedPCIDevice),
        default={},
    )
```
In this case, you may be absolutely sure that:
1. There is always RestrictedPCIDevices.driver_names and RestrictedPCIDevices.pci_ids exists
2. The values of this field can be either empty dict (nullable) or Dict[str, RestrictedPCIDevice]

Reply:
Just for public reference, it was decided that Dict will not be used. so models need to be updated to lists.